### PR TITLE
Scoring & estimator (Sabaki port)

### PR DESCRIPTION
**Pull request by mbarc11**
## Summary

Adds territory/area scoring, an influence-based score estimator, and a Monte-Carlo dead-stone detector — all ported from Sabaki's open-source modules into pure Dart, with a click-to-mark UX that mirrors Sabaki's scoring mode.

Everything lives in a new `lib/src/scoring.dart` (exported from `lib/golo.dart`). 127 tests pass, including a fixture-pinned regression test that verifies the dead-stone port matches Sabaki's published reference output.

### Strict scoring (Japanese / Chinese)
Direct port of [`@sabaki/influence/areaMap.js`](https://github.com/SabakiHQ/influence) and Sabaki's `getScore` in `src/modules/utils.js`.

```dart
final s = Scorer.finalScore(board, komi: 6.5, handicap: 0);
s.areaScore       // Chinese
s.territoryScore  // Japanese
```

### Influence estimator (Bouzy-style)
Direct port of `@sabaki/influence/influenceMap.js` (areaMap → nearest-neighbor + radiance maps → postprocessing).

```dart
final s = Scorer.estimateScore(board, komi: 6.5);
final overlay = Scorer.estimate(board); // {-1, 0, 1} sign-map
```

### Monte-Carlo dead-stone detection
Direct port of [`@sabaki/deadstones`](https://github.com/SabakiHQ/deadstones) (Rust → Dart): `PseudoBoard`, `getFloatingStones` static prefilter, `playTillEnd` random playouts, `getProbabilityMap`, related-chains refinement.

```dart
final dead = Scorer.guessDeadStones(board, iterations: 100, seed: 42);
final score = Scorer.scoreEndgame(board, komi: 6.5);
```

Pinned to Sabaki's own test fixture: `Scorer.findFloatingStones matches Sabaki @sabaki/deadstones finished-board fixture` produces the exact 11-vertex output published in Sabaki's tests.

### Click-to-mark UX (Sabaki scoring mode)
Mirrors Sabaki's UI: a click on a stone toggles its **whole group** — chain + same-colour chains reachable through empty cells (`getRelatedChains`).

```dart
var dead = <Vertex>{};
dead = Scorer.toggleDeadGroup(dead, board, clickedVertex);
final score = Scorer.scoreWithDeadStones(board, dead, komi: 6.5);
```

`Scorer.deadStoneSelection` returns the group that would toggle (without applying it), useful for hover/preview UI. Pass `includeRelatedChains: false` for Sabaki's estimator-mode behaviour (just the clicked chain).

### Example CLI
`example/score_sgf.dart` loads any SGF and prints strict, estimator, and MCTS scores side by side. With `--interactive` it drops into a REPL where typing a coordinate (e.g. `D4`) toggles a group's dead/alive state and re-displays the score; `auto` re-runs the MCTS suggestion, `reset` clears the marks.

```
dart run example/score_sgf.dart path/to/game.sgf
dart run example/score_sgf.dart path/to/game.sgf --interactive
```

## API surface

```dart
// Sign-map helpers (low-level)
SignMap signMapOf(Board)
SignMap areaMap(SignMap)
SignMap influenceMap(SignMap, {discrete, maxDistance, minRadiance})
Score   score(Board, SignMap, {komi, handicap})

// Board-level facade
Scorer.boardAreaMap(board)
Scorer.estimate(board)
Scorer.finalScore(board, {komi, handicap})
Scorer.estimateScore(board, {komi, handicap})
Scorer.scoreEndgame(board, {komi, handicap, iterations, seed})
Scorer.guessDeadStones(board, {iterations, finished, seed})
Scorer.staticGuessDeadStones(board)
Scorer.findFloatingStones(board)
Scorer.settleForScoring(board, {iterations, seed})

// Click-to-mark
Scorer.deadStoneSelection(board, vertex, {includeRelatedChains})
Scorer.toggleDeadGroup(deadSet, board, vertex, {includeRelatedChains})
Scorer.scoreWithDeadStones(board, deadStones, {komi, handicap})
```

## Tests

- 127 total (108 existing + 19 new).
- Coverage: area maps and dame, captures arithmetic, estimator output shape & corner classification, MCTS determinism + floating-stones superset property, click expansion + toggle semantics, score-with-dead-stones bookkeeping.
- The Sabaki-fixture regression test ensures the MCTS port doesn't drift from upstream behaviour.

## Test plan

- [x] `dart analyze` — clean
- [x] `dart test` — 127/127 pass
- [x] `dart run example/score_sgf.dart example/sgf/test1.sgf` — produces strict / estimator / MCTS scores
- [x] `dart run example/score_sgf.dart example/sgf/test1.sgf -i` — interactive REPL toggles correctly

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by ReiRev** [comment]
Thank you. Could you also update license file?

**Comment by mbarc11** [comment]
Done — pushed [`48c699c`](https://github.com/ReiRev/dart_golo/pull/35/commits/48c699c). The LICENSE now includes MIT notices for the three Sabaki sources this PR is derived from, following the existing pattern used for `@sabaki/go-board` and `SabakiHQ/sgf`:

- **`@sabaki/influence`** — for the area-map and influence estimator (`lib/src/scoring.dart`)
- **`@sabaki/deadstones`** — for the Monte-Carlo dead-stone detector (`lib/src/scoring.dart`)
- **`Sabaki`** — for the `getScore` territory/area arithmetic and the click-to-mark scoring UX (`lib/src/scoring.dart`)

All four (including upstream's existing `@sabaki/go-board` and `SabakiHQ/sgf`) are MIT-licensed and copyright Yichuan Shen.

**Comment by mbarc11** [comment]
Thanks for the review — all four points addressed in [`caffbb2`](https://github.com/ReiRev/dart_golo/pull/35/commits/caffbb2).

**1. `_radianceMap.castRadiance` — `removeAt(0)` is O(n) per pop.** Fixed. Switched to `dart:collection`'s `ListQueue` with `addLast`/`removeFirst` (O(1) per pop). This is the hot loop for every estimator call.

**2. `scoreEndgame` redundant clone.** Fixed. The `settled.cleaned` from `settleForScoring` was being thrown away while `scoreWithDeadStones` re-cloned the board internally. Now calls `guessDeadStones` + `scoreWithDeadStones` directly, saving one clone+remove pass per call.

**3. `makePseudoMove` doc/code mismatch.** The actual behaviour is correct (and matches `pseudo_board.rs::make_pseudo_move`) — off-board neighbours are filtered out by `getNeighbors`, so the predicate's null check was dead code, and the rule it implements is "no in-bounds neighbour is empty or opponent" (own colour or wall). Updated the doc to describe that accurately, simplified the predicate to drop the dead `s == null` check, and added an inline note explaining the off-board / wall equivalence.

**4. `minRadiance` rounding.** This is intentional and matches `Math.round(...) < minRadiance` in `@sabaki/influence/src/influenceMap.js`. Removing the `round()` would diverge from upstream and shift the boundary (e.g. `rad = 1.9` would suddenly fail at `minRadiance: 2`). Added an inline comment at the call site flagging that the rounding is verbatim, and clarified the doc on the parameter so callers know the comparison is against `rad.round()`, not `rad`.

`dart analyze` clean, full suite (127 tests) still passing.

File: lib/src/scoring.dart [hunk 1]
```diff
@@ -0,0 +1,1138 @@
...
+    s == null ? 0 : (s == Stone.black ? 1 : -1);
+
+/// Builds a sign-map from a [Board].
+SignMap signMapOf(Board board) {
+  return List.generate(
+    board.height,
+    (y) => List.generate(board.width, (x) => _signOf(board.get((x: x, y: y)))),
+  );
+}
+
```
Comment (ReiRev) [inline lib/src/scoring.dart:23]: I know that Sabaki uses this kind of signed map for the board state, but it is not Dart-like, so I intentionally chose not to follow that style in this implementation. This representation should be centralized in board.dart, not declared here. Please use the Board class from board.dart. This applies to all related lines.

File: LICENSE [hunk 2]
```diff
@@ -81,3 +81,96 @@ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
+
```
Comment (ReiRev) [inline LICENSE:84]: Please also update README.md

File: lib/src/scoring.dart [hunk 3]
```diff
@@ -0,0 +1,1138 @@
...
+  return list.reduce((a, b) => a + b) / list.length;
+}
+
+/// Computes an influence-based estimator map.
+///
+/// Port of `@sabaki/influence/src/influenceMap.js`. Returns a map where the
+/// sign of each entry indicates the predicted owner: `+1` Black, `-1` White,
+/// `0` neutral. Stones keep their literal sign.
+///
+/// - [discrete]: when true (default for the estimator), output is snapped to
```
Comment (ReiRev) [inline lib/src/scoring.dart:277]: This API description is wrong.

Comment (ReiRev) [inline lib/src/scoring.dart:277]: This API description is wrong.

File: lib/src/scoring.dart [hunk 4]
```diff
@@ -0,0 +1,1138 @@
...
+    }
+  }
+
+  return map;
+}
+
+const int _unset = -1000000;
+
+/// Connected component of empty points starting at (x, y).
+List<List<int>> _emptyChain(SignMap signMap, int x, int y) {
```
Comment (ReiRev) [inline lib/src/scoring.dart:88]: Please do not return List<List<int>>. This will increase the possibility of bugs.

File: lib/src/scoring.dart [hunk 5]
```diff
@@ -0,0 +1,1138 @@
...
+
+  return map;
+}
+
+/// Radiance (influence strength) emitted by each [sign]-coloured group.
+///
+/// Port of `@sabaki/influence/src/radianceMap.js`. Each chain of stones casts
+/// a BFS wave outward up to depth [p1]; reflections off the wall add [p3];
+/// non-mirrored cells receive `p2 / (d / p1 * 6 + 1)`.
+List<List<double>> _radianceMap(SignMap data, int sign,
```
Comment (ReiRev) [inline lib/src/scoring.dart:176]: This is no longer just a few local helpers. The module’s core logic is effectively outside the owning type, which makes the structure harder to follow and evolve. Please put this function inside the class. This applies to all other related functions.

File: lib/src/scoring.dart [hunk 6]
```diff
@@ -0,0 +1,1138 @@
...
+//
+// Direct port of Sabaki's `@sabaki/deadstones` (Rust → WASM):
+//   - [_PseudoBoard] mirrors `PseudoBoard` in src/pseudo_board.rs
+//   - [_playTillEnd] mirrors `play_till_end` in src/deadstones.rs
+//   - [_mctsGuessDeadStones] mirrors the `guess` function: a static
+//     `getFloatingStones` prefilter plus N random playouts, then a
+//     related-chains refinement pass.
+// ---------------------------------------------------------------------------
+
+class _PseudoBoard {
```
Comment (ReiRev) [inline lib/src/scoring.dart:778]: _PseudoBoard is a second internal board model used only by the Monte Carlo dead-stone estimator. It reimplements core board operations outside the existing Board type, which increases duplication and makes the scoring pipeline depend on a parallel rules engine rather than the library’s canonical board abstraction.

**Review by ReiRev — CHANGES_REQUESTED** [review]
I reviewed this PR in detail, and I do not think it is mergeable in its current state.

The main problem is not just individual bugs. The implementation appears to be a direct port of the JavaScript/Sabaki code without enough understanding of the existing Dart codebase and its architecture. As a result, this PR introduces a lot of duplicated logic that already exists in `Board` or related classes, and that duplication makes the code harder to maintain and increases the risk of subtle bugs.

If AI is used to implement this kind of large feature, please first instruct it to understand the overall architecture of this repository and keep the implementation consistent with the existing abstractions. After the initial implementation, please also run additional review/refactoring prompts focused specifically on:

- architectural consistency with the existing codebase
- removal of duplicated logic
- correctness of the port
- edge cases and test coverage

Thank you.
